File: Problems/0397-integer-replacement.py
class Solution(object):
    def integerReplacement(self, n):
        """
        :type n: int
        :rtype: int
        """
        ans = 0

        while n != 1:
            if n & 1 == 0:
                n >>= 1
            elif n == 3 or (n & 3) == 1:
                n -= 1
            else:
                n += 1
            ans += 1

        return ans

File: Problems/0397-integer-replacement.cpp
#include <bits/stdc++.h>
using namespace std;

// LeetCode solution starts from here

class Solution {
public:
    int integerReplacement(int n) {
        long long x = n;
        int ans = 0;

        while (x != 1) {
            if ((x & 1) == 0) {
                x >>= 1;
            } else if (x == 3 || (x & 3) == 1) {
                x--;
            } else {
                x++;
            }
            ans++;
        }

        return ans;
    }
};
